### PR TITLE
Add unit tests for FindCommand class #110

### LINKED ISSUE #110: Add unit tests for FindCommand class
as per title

### PR DESCRIPTION
**Pull request by KnewYouWereTrouble**
Fixes #110
Ready for review.

---
This change is [Reviewable](https://reviewable.io/reviews/se-edu/addressbook-level2/120)

### CODE REVIEW

File: src/seedu/addressbook/data/AddressBook.java [hunk 1]
```diff
@@ -143,4 +143,12 @@ public UniquePersonList getAllPersons() {
     public UniqueTagList getAllTags() {
         return new UniqueTagList(allTags);
     }
+    
+    @Override
+    public boolean equals(Object other) {
+        return other == this // short circuit if same object
+                || (other instanceof AddressBook // instanceof handles nulls
+                && this.allPersons.equals(((AddressBook) other).allPersons)
+                && this.allTags.equals(((AddressBook) other).allTags));
```
Comment (damithc) [inline src/seedu/addressbook/data/AddressBook.java]: These two lines should be indented further because it is misleading to align `||` with `&&` (they are not at the same level in the condition tree)

File: test/java/seedu/addressbook/commands/FindCommandTest.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+import seedu.addressbook.data.person.Person;
+import seedu.addressbook.testutil.TypicalTestPersons;
+
+/**
+ * Test class for the Find Command
+ */
+public class FindCommandTest extends CommandTest {
+
+    @Test
+    public void execute_find_isCaseSensitive() throws IllegalValueException {
```
Comment (damithc) [inline test/java/seedu/addressbook/commands/FindCommandTest.java]: This name does not follow the convention. For example, the 3rd part of the name should be the expected result.

File: test/java/seedu/addressbook/testutil/TestMain.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+import seedu.addressbook.parser.Parser;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * Test class for Main
+ */
+public class TestMain {
```
Comment (damithc) [inline test/java/seedu/addressbook/testutil/TestMain.java]: This PR is for testing the `FindCommand` class, not the '`find` command' as typed by the user. Hence, I don't see the need for this class. Compare with #118

File: test/java/seedu/addressbook/testutil/TypicalTestPersons.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+            jobs = new PersonBuilder().withName("Steve Jobs").withPhone("91934267", true)
+                    .withEmail("sj@gmail.com", true).withAddress("3 Clementi Road", false).build();
+
+            donald = new PersonBuilder().withName("Donald Knuth").withPhone("91214567", true)
+                    .withEmail("dk@gmail.com", false).withAddress("4 Clementi Road", true).build();
+
+            hopper = new PersonBuilder().withName("Grace Hopper").withPhone("83434567", true)
+                    .withEmail("gh@gmail.com", true).withAddress("5 Clementi Road", true).build();
+
```
Comment (damithc) [inline test/java/seedu/addressbook/testutil/TypicalTestPersons.java]: When defining test data, try to follow a pattern that is easy to remember. For example, the first person's name has initials A. B.  (e.g. Alice Betsy), second person's name has initials B. C. (Ben Charles) and so on. When things always follow a pattern, anything going out of that pattern (e.g. Ben suddenly appears at the start of the list) gets noticed immediately and sometimes can lead to discover bugs.

File: test/java/seedu/addressbook/testutil/PersonBuilder.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    public PersonBuilder() {
+        this.person = new TestPerson();
+    }
+
+    public PersonBuilder withName(String name) throws IllegalValueException {
+        this.person.setName(new Name(name));
+        return this;
+    }
+
+    public PersonBuilder withPhone(String phone, boolean isPrivate) throws IllegalValueException {
```
Comment (damithc) [inline test/java/seedu/addressbook/testutil/PersonBuilder.java]: While we are at it, we can also have a `withPrivatePhone` method.

Comment (KnewYouWereTrouble) [inline test/java/seedu/addressbook/testutil/PersonBuilder.java]: Okay. Implemented two methods for each field. 
For example withPublicPhone and withPrivatePhone.

File: test/java/seedu/addressbook/testutil/TestMain.java (hunk 3 above)
Comment (KnewYouWereTrouble) [inline test/java/seedu/addressbook/testutil/TestMain.java]: Okay removed!

File: test/java/seedu/addressbook/commands/FindCommandTest.java (hunk 2 above)
Comment (KnewYouWereTrouble) [inline test/java/seedu/addressbook/commands/FindCommandTest.java]: Okay changed to execute_find_matchesOnlyIfCaseSensitive

**Comment by damithc** [comment]
Seems like you have unrelated commits in this branch.

**Comment by KnewYouWereTrouble** [comment]
Sorry, I rebased the commits wrongly! Have rectified.

File: test/java/seedu/addressbook/commands/FindCommandTest.java [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+    private AddressBook addressBook;
+    private AddressBook unmutatedAddressBook;
+
+    @Before public void setUp() {
+        this.addressBook = new TypicalTestPersons().getTypicalAddressBook();
+        this.unmutatedAddressBook = new TypicalTestPersons().getTypicalAddressBook();
+    }
+
+    @Test
+    public void execute_find_matchesOnlyIfCaseSensitive() throws IllegalValueException {
```
Comment (damithc) [inline test/java/seedu/addressbook/commands/FindCommandTest.java]: As we are testing `FindCommand`, `execute` implies `find`.
The method name can be `execute_sameWordDifferentCase_notMatched()`. In that case, this method needs to be split into two. Alternatively, you can name it `execute_caseSensitivity()` and test all cases of case sensitivity within the same method. In that case, use comments inside the method to describe each case.

File: test/java/seedu/addressbook/commands/FindCommandTest.java [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+        List<ReadOnlyPerson> expectedOne = TestUtil.createList(new Person(TypicalTestPersons.amy));
+        List<ReadOnlyPerson> expectedZero = Collections.emptyList();
```
Comment (damithc) [inline test/java/seedu/addressbook/commands/FindCommandTest.java]: Define the variable closer to where it will be used.

File: test/java/seedu/addressbook/commands/FindCommandTest.java [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+        
+        FindCommand command = createFindCommand(caseSensitiveKeywords);
+        assertCommandBehavior(command, Command.getMessageForPersonListShownSummary(expectedOne),
+                unmutatedAddressBook, addressBook);
+        
+        command = createFindCommand(caseInsensitiveKeywords);
+        assertCommandBehavior(command, Command.getMessageForPersonListShownSummary(expectedZero),
+                unmutatedAddressBook, addressBook);
+
+
```
Comment (damithc) [inline test/java/seedu/addressbook/commands/FindCommandTest.java]: Why two blank lines?

File: test/java/seedu/addressbook/commands/FindCommandTest.java [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+        this.addressBook = new TypicalTestPersons().getTypicalAddressBook();
+        this.unmutatedAddressBook = new TypicalTestPersons().getTypicalAddressBook();
+    }
+
+    @Test
+    public void execute_find_matchesOnlyIfCaseSensitive() throws IllegalValueException {
+        List<ReadOnlyPerson> expectedOne = TestUtil.createList(new Person(TypicalTestPersons.amy));
+        List<ReadOnlyPerson> expectedZero = Collections.emptyList();
+
+        String[] caseSensitiveKeywords = {"Amy"};
```
Comment (damithc) [inline test/java/seedu/addressbook/commands/FindCommandTest.java]: The name doesn't do a very good job of describing what the variable is. A word by itself is not 'case sensitive'. `keyWordsWithMatchingCase`?

File: test/java/seedu/addressbook/commands/FindCommandTest.java (hunk 7 above)
Comment (damithc) [inline test/java/seedu/addressbook/commands/FindCommandTest.java]: You can use the same variable `expected` instead of declaring two variables with somewhat 'odd' names. :-)

File: test/java/seedu/addressbook/util/TestPerson.java [hunk 10]
```diff
@@ -0,0 +1,59 @@
+package seedu.addressbook.util;
+
+import seedu.addressbook.data.person.Address;
+import seedu.addressbook.data.person.Email;
+import seedu.addressbook.data.person.Name;
+import seedu.addressbook.data.person.Phone;
+import seedu.addressbook.data.person.ReadOnlyPerson;
+import seedu.addressbook.data.tag.UniqueTagList;
+
+public class TestPerson implements ReadOnlyPerson {
```
Comment (damithc) [inline test/java/seedu/addressbook/util/TestPerson.java]: Why do we need this class? In any case, there should be a header comment to explain what it is.

File: test/java/seedu/addressbook/util/TypicalTestPersons.java [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+ * Class to generate typical test persons
+ */
+public class TypicalTestPersons {
+
+    public static TestPerson amy, bill, candy, donald, grace;
+
+    public TypicalTestPersons() {
+        try {
+            amy = new PersonBuilder().withName("Amy Buck").withPublicPhone("91234567")
+                    .withPublicEmail("cs@gmail.com").withPublicAddress("1 Clementi Road").build();
```
Comment (damithc) [inline test/java/seedu/addressbook/util/TypicalTestPersons.java]: telephone `91119111` and email `ab@gmail.com` instead of random values? (for the same reasons as before; if we follow patterns, it's easy to detect when things go wrong)

File: test/java/seedu/addressbook/commands/FindCommandTest.java (hunk 6 above)
Comment (KnewYouWereTrouble) [inline test/java/seedu/addressbook/commands/FindCommandTest.java]: Okay! I have split the original method into 2.

File: test/java/seedu/addressbook/util/TestPerson.java (hunk 10 above)
Comment (KnewYouWereTrouble) [inline test/java/seedu/addressbook/util/TestPerson.java]: Okay I removed this class. PersonBuilder now builds an instance of the Person class.

**Comment by KnewYouWereTrouble** [comment]
Implemented the requested changes!

File: src/seedu/addressbook/data/person/UniquePersonList.java [hunk 12]
```diff
@@ -124,11 +124,12 @@ public void clear() {
     public Iterator<Person> iterator() {
         return internalList.iterator();
     }
-
+
```
Comment (damithc) [inline src/seedu/addressbook/data/person/UniquePersonList.java]: Blank lines should not contain spaces.

File: test/java/seedu/addressbook/commands/FindCommandTest.java [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+        
+        String[] keywordsWithNonMatchingCase = {"aMy"};
+        
+        FindCommand command = createFindCommand(keywordsWithNonMatchingCase);
+        assertCommandBehavior(command, Command.getMessageForPersonListShownSummary(expected),
+                unmutatedAddressBook, addressBook);
+    }
+    
+    
+
```
Comment (damithc) [inline test/java/seedu/addressbook/commands/FindCommandTest.java]: Why 3 blank lines here?

File: test/java/seedu/addressbook/commands/FindCommandTest.java [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+    @Test
+    public void execute_find_matchesOnlyFullWordsInNames() throws IllegalValueException {
```
Comment (damithc) [inline test/java/seedu/addressbook/commands/FindCommandTest.java]: Renames this method and subsequent methods to to be consistent with the previous two methods.

File: test/java/seedu/addressbook/commands/FindCommandTest.java [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+        List<ReadOnlyPerson> expected = Collections.emptyList();
+        
+        String[] keywords = {"my"};
+        
+        FindCommand command = createFindCommand(keywords);
+        assertCommandBehavior(command, Command.getMessageForPersonListShownSummary(expected),
+                unmutatedAddressBook, addressBook);
```
Comment (damithc) [inline test/java/seedu/addressbook/commands/FindCommandTest.java]: Increasing indentation like this can improve readability (the opening and closing brackets appear on extreme left and extreme right respectively of the parameter text)
```
assertCommandBehavior(command, Command.getMessageForPersonListShownSummary(expected),
                                                          unmutatedAddressBook, addressBook);
```

File: test/java/seedu/addressbook/commands/FindCommandTest.java [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+
+        CommandResult result = command.execute();
+        
+        assertEquals(expectedMessage, result.feedbackToUser);
+        assertEquals(expectedAddressBook, actualAddressBook);
+    }
+
+    
+    /**
+     * Creates a new find command.
```
Comment (damithc) [inline test/java/seedu/addressbook/commands/FindCommandTest.java]: `FindCommand`

Comment (damithc) [inline test/java/seedu/addressbook/commands/FindCommandTest.java]: Anyway, can be omitted as it doesn't add any value. It's the same as the method name.

File: test/java/seedu/addressbook/util/TypicalPersons.java [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+            grace = new PersonBuilder().withName("Grace Hopper").withPublicPhone("95559555")
+                    .withPrivateEmail("gh@gmail.com").withPrivateAddress("5 Clementi Road").build();
+
+        } catch (IllegalValueException e) {
+            e.printStackTrace();
+            assert false : "not possible";
+        }
+    }
+
+    public void loadAddressBookWithSampleData(AddressBook ab) {
```
Comment (damithc) [inline test/java/seedu/addressbook/util/TypicalPersons.java]: does this need to be public?

**Comment by damithc** [comment]
@KnewYouWereTrouble what's the status of this PR?

**Comment by KnewYouWereTrouble** [comment]
Ready for review!

File: test/java/seedu/addressbook/commands/FindCommandTest.java [hunk 18]
```diff
@@ -0,0 +1,97 @@
...
+
+    @Test
+    public void execute_sameWordSameCase_matched() throws IllegalValueException {
+        List<ReadOnlyPerson> expected = TestUtil.createList(new Person(TypicalPersons.amy));
+        
+        String[] keywordsWithMatchingCase = {"Amy"};
+        
+        FindCommand command = createFindCommand(keywordsWithMatchingCase);
+        assertCommandBehavior(command, Command.getMessageForPersonListShownSummary(expected),
+                                                                    unmutatedAddressBook, addressBook);
```
Comment (damithc) [inline test/java/seedu/addressbook/commands/FindCommandTest.java]: `Command.getMessageForPersonListShownSummary` seems to be repeated every time. May be you can put that part inside the `assertCommandBehavior` method itself?

File: test/java/seedu/addressbook/commands/FindCommandTest.java [hunk 19]
```diff
@@ -0,0 +1,97 @@
...
+    }
+    
+    @Test
+    public void execute_sameWordDifferentCase_notMatched() throws IllegalValueException {
+        List<ReadOnlyPerson> expected = Collections.emptyList();
```
Comment (damithc) [inline test/java/seedu/addressbook/commands/FindCommandTest.java]: Declare this as a constant `EMPTY_LIST` and use it everywhere?

Comment (damithc) [inline test/java/seedu/addressbook/commands/FindCommandTest.java]: Alternatively, you can create this utility method:
```java
public void assertNoResults(...){
    assertCommandBehavior(..., Collections.emptyList(), ...);
}
```
In that case you may want to rename `assertCommandBehavior` as `assertResults` (for symmetry)

File: test/java/seedu/addressbook/commands/FindCommandTest.java (hunk 18 above)
Comment (damithc) [inline test/java/seedu/addressbook/commands/FindCommandTest.java]: Same thing for `unmutatedAddressBook`. It can be inside the method itself instead of passing as a parameter every time?

File: test/java/seedu/addressbook/commands/FindCommandTest.java [hunk 20]
```diff
@@ -0,0 +1,97 @@
...
+        assertCommandBehavior(command, Command.getMessageForPersonListShownSummary(expected),
+                                                                    unmutatedAddressBook, addressBook);
+    }
+
+    @Test
+    public void execute_multipleKeywords_matched() throws IllegalValueException {
+        List<ReadOnlyPerson> expected = TestUtil.createList(new Person(TypicalPersons.amy),
+                                                            new Person(TypicalPersons.bill));
+        
+        String[] keywords = {"Amy", "Bill"};
```
Comment (damithc) [inline test/java/seedu/addressbook/commands/FindCommandTest.java]: You seem to be testing the first name only. What if there is a bug that limits the string matching to the first word of the name only?

File: test/java/seedu/addressbook/util/TypicalPersons.java [hunk 21]
```diff
@@ -0,0 +1,57 @@
...
+
+        } catch (IllegalValueException e) {
+            e.printStackTrace();
+            assert false : "not possible";
+        }
+    }
+
+    private void loadAddressBookWithSampleData(AddressBook ab) {
+        try {
+            for (Person t : getTypicalPersons()) {
```
Comment (damithc) [inline test/java/seedu/addressbook/util/TypicalPersons.java]: Why `t` and not `p`?

File: test/java/seedu/addressbook/commands/FindCommandTest.java (hunk 18 above)
Comment (KnewYouWereTrouble) [inline test/java/seedu/addressbook/commands/FindCommandTest.java]: Okay! I moved them into the `assertCommandBehavior` method!

File: test/java/seedu/addressbook/commands/FindCommandTest.java (hunk 20 above)
Comment (KnewYouWereTrouble) [inline test/java/seedu/addressbook/commands/FindCommandTest.java]: Okay! added another keyword that is the last name of the person

File: test/java/seedu/addressbook/util/TypicalPersons.java (hunk 21 above)
Comment (KnewYouWereTrouble) [inline test/java/seedu/addressbook/util/TypicalPersons.java]: Edited!

**Comment by KnewYouWereTrouble** [comment]
Ready for review!

File: src/seedu/addressbook/data/person/UniquePersonList.java [hunk 22]
```diff
@@ -131,4 +131,5 @@ public boolean equals(Object other) {
                 || (other instanceof UniquePersonList // instanceof handles nulls
                         && this.internalList.equals(((UniquePersonList) other).internalList));
     }
+
```
Comment (damithc) [inline src/seedu/addressbook/data/person/UniquePersonList.java]: Unrelated change?

File: test/java/seedu/addressbook/commands/FindCommandTest.java [hunk 23]
```diff
@@ -0,0 +1,89 @@
...
+    }
+
+    @Test
+    public void execute_sameWordSameCase_matched() throws IllegalValueException {
+        List<ReadOnlyPerson> expected = TestUtil.createList(new Person(TypicalPersons.amy));
+        
+        String[] keywordsWithMatchingCase = {"Amy"};
+        
+        FindCommand command = createFindCommand(keywordsWithMatchingCase);
+        assertCommandBehavior(command, expected, addressBook);
```
Comment (damithc) [inline test/java/seedu/addressbook/commands/FindCommandTest.java]: Still too verbose. Out of this 4 lines, the only things that are specific to this test case is `Amy` and `amy`. Everything else is just boilerplate and can be pushed into the utility method.

File: test/java/seedu/addressbook/commands/FindCommandTest.java [hunk 24]
```diff
@@ -0,0 +1,89 @@
...
+/**
+ * Test class for the Find Command
+ */
+public class FindCommandTest {
+
+    private AddressBook addressBook;
+    
+    private final List<ReadOnlyPerson> EMPTY_LIST = Collections.emptyList();
+
+    @Before public void setUp() {
```
Comment (damithc) [inline test/java/seedu/addressbook/commands/FindCommandTest.java]: Annotation should be in a separate line.

File: test/java/seedu/addressbook/util/TypicalPersons.java [hunk 25]
```diff
@@ -0,0 +1,57 @@
...
+import seedu.addressbook.data.exception.IllegalValueException;
+import seedu.addressbook.data.person.Person;
+import seedu.addressbook.data.person.UniquePersonList;
+
+/**
+ * Class to generate typical test persons
+ */
+public class TypicalPersons {
+
+    public static Person amy, bill, candy, donald, grace;
```
Comment (pyokagan) [inline test/java/seedu/addressbook/util/TypicalPersons.java]: @damithc Addressbook-Level4 has the same problem too. I'll open up an issue there.

Comment (pyokagan) [inline test/java/seedu/addressbook/util/TypicalPersons.java]: Shouldn't we just copy over the code from AddressBook-Level4 in order to share as much code as possible?

In AddressBook-Level4 we have alice, benson, carl etc. but here we have amy, bill, candy etc. Seems like it will cause trouble when we want to share tests across different levels in the future.

Comment (damithc) [inline test/java/seedu/addressbook/util/TypicalPersons.java]: @pyokagan  I agree. But I don't mind doing it as a separate issue. The extra work caused by duplication of code is a good source of 'low hanging fruit' issues for new contributors to do. 

On a somewhat related note, at the moment I'm not too eager to unify similarities between the levels. At our final steady state it may be desirable to have different levels doing the same thing in different ways, so that students get to see more ways of doing the same thing. As we are not yet certain what our final steady state is going to be, I don't mind if levels evolved somewhat independently for now.

In hindsight, I'm not sure if we should use the Builder pattern in level 2. It would be good for one level (level 2?) to not use the Builder pattern so that it can be contrasted with the next level (level 3?) that improves the situation by applying the Builder pattern. Anyway, that's something we can consider on a future day. We can leave it be for now. :-p

Comment (pyokagan) [inline test/java/seedu/addressbook/util/TypicalPersons.java]: @damithc says:
> The extra work caused by duplication of code is a good source of 'low hanging fruit' issues for new contributors to do.

It's more work for reviewers through, and will slow down the project. Reviewers may not be happy that they have to review the same things over and over again, but subtly different each time due to differences in code bases.

> at the moment I'm not too eager to unify similarities between the levels.

I understand. However I hope that until there is a clear direction on how the different levels are going to differ from each other, we try to prevent them from diverging unnecessarily, simply because it is *much* easier to make code bases diverge rather than converge.

For an example of "diverging unnecessarily", for this PR I hope that the test data is kept exactly the same as the other levels, because future tests will be written *for* the test data. Having test data that is different will just cause unnecessary friction in the future should we ever try to bring tests from one level to another. Just try to cover our bases here :-)

**Comment by damithc** [comment]
@KnewYouWereTrouble reminder

**Comment by KnewYouWereTrouble** [comment]
Ready for review!

File: test/java/seedu/addressbook/util/PersonBuilder.java [hunk 26]
```diff
@@ -0,0 +1,68 @@
...
+import seedu.addressbook.data.person.Email;
+import seedu.addressbook.data.person.Name;
+import seedu.addressbook.data.person.Person;
+import seedu.addressbook.data.person.Phone;
+import seedu.addressbook.data.tag.UniqueTagList;
+
+/**
+ * Factory class for building TestPerson
+ */
+public class PersonBuilder {
```
Comment (damithc) [inline test/java/seedu/addressbook/util/PersonBuilder.java]: I think we can do without a PersonBuilder. Level 2 a bit too early to do this kind of thing. Sorry I didn't think about it earlier.

File: test/java/seedu/addressbook/commands/FindCommandTest.java [hunk 27]
```diff
@@ -0,0 +1,84 @@
...
+        this.addressBook = new TypicalPersons().getTypicalAddressBook();
+    }
+
+    @Test
+    public void execute_sameWordSameCase_matched() throws IllegalValueException {
+        List<ReadOnlyPerson> expected = TestUtil.createList(new Person(TypicalPersons.amy()));
+        String[] keywordsWithMatchingCase = {"Amy"};
+        
+        assertFindCommandBehavior(keywordsWithMatchingCase, expected, addressBook);
+    }
```
Comment (damithc) [inline test/java/seedu/addressbook/commands/FindCommandTest.java]: Can be made even more concise:
``` java
public void execute() throws exception {
    //same word, same case: matched
    assertFindCommandBehavior( {"Amy"}, TypicalPersons.amy()); 

    //same word, different case: not matched
    assertFindCommandBehavior( {"amy"});

   ...   
 }

private void assertFindCommandBehavior(String[] keywords, Person...  expectedPersonList){
    ...
}
```

Comment (damithc) [inline test/java/seedu/addressbook/commands/FindCommandTest.java]: Or even this (slightly better):
``` java
public void execute() throws exception {
    //same word, same case: matched
    assertFindCommandBehavior( {"Amy"}, {TypicalPersons.amy()}); 

    //same word, different case: not matched
    assertFindCommandBehavior( {"amy"}, {});

   ...   
 }

private void assertFindCommandBehavior(String[] keywords, Person[]  expectedPersons){
    ...
}
```

**Comment by KnewYouWereTrouble** [comment]
Ready for review

File: test/java/seedu/addressbook/commands/FindCommandTest.java [hunk 28]
```diff
@@ -0,0 +1,68 @@
...
+        assertFindCommandBehavior(new String[]{"Amy"}, TestUtil.createList(td.amy));
+        
+        //same word, different case: not matched
+        assertFindCommandBehavior(new String[]{"aMy"}, EMPTY_LIST);
+        
+        //partial word: not matched
+        assertFindCommandBehavior(new String[]{"my"}, EMPTY_LIST);
+        
+        //multiple words: matched
+        assertFindCommandBehavior(new String[]{"Amy", "Bill", "Destiny"}, TestUtil.createList(td.amy, td.bill, td.candy));
```
Comment (damithc) [inline test/java/seedu/addressbook/commands/FindCommandTest.java]: You have a case of matching first name only, and second name only. How about a case of matching both names? i.e. `"Amy", "Bill", "Candy", "Destiny"`

File: test/java/seedu/addressbook/commands/FindCommandTest.java [hunk 29]
```diff
@@ -0,0 +1,68 @@
...
+    }
+
+
```
Comment (damithc) [inline test/java/seedu/addressbook/commands/FindCommandTest.java]: Extra blank line

File: test/java/seedu/addressbook/commands/FindCommandTest.java [hunk 30]
```diff
@@ -0,0 +1,68 @@
...
+    /**
+     * Executes the find command, and checks that the execution was what we had expected.
```
Comment (damithc) [inline test/java/seedu/addressbook/commands/FindCommandTest.java]: Can be more specific. e.g. `Executes the find command for the given keywords and verifies the result matches the persons in the expectedPersonList exactly`

File: test/java/seedu/addressbook/commands/FindCommandTest.java [hunk 31]
```diff
@@ -0,0 +1,68 @@
...
+        
+        //same word, different case: not matched
+        assertFindCommandBehavior(new String[]{"aMy"}, EMPTY_LIST);
+        
+        //partial word: not matched
+        assertFindCommandBehavior(new String[]{"my"}, EMPTY_LIST);
+        
+        //multiple words: matched
+        assertFindCommandBehavior(new String[]{"Amy", "Bill", "Destiny"}, TestUtil.createList(td.amy, td.bill, td.candy));
+    }
```
Comment (damithc) [inline test/java/seedu/addressbook/commands/FindCommandTest.java]: Other interesting cases to test:
* Repeated keyword
* Keyword matching a word in address

File: test/java/seedu/addressbook/commands/FindCommandTest.java [hunk 32]
```diff
@@ -0,0 +1,68 @@
...
+
+import seedu.addressbook.data.AddressBook;
+import seedu.addressbook.data.exception.IllegalValueException;
+import seedu.addressbook.data.person.ReadOnlyPerson;
+import seedu.addressbook.util.TestUtil;
+import seedu.addressbook.util.TypicalPersons;
+
+/**
+ * Test class for the Find Command
+ */
```
Comment (damithc) [inline test/java/seedu/addressbook/commands/FindCommandTest.java]: This header comment doesn't add any value. Can be removed.

**Comment by damithc** [comment]
Reviewed 1 of 15 files at r1, 4 of 14 files at r2, 1 of 5 files at r4, 3 of 4 files at r9.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/se-edu/addressbook-level2/120)*

**Comment by KnewYouWereTrouble** [comment]
Ready for review!

**Comment by damithc** [comment]
Reviewed 2 of 2 files at r11.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[src/seedu/addressbook/data/tag/UniqueTagList.java, line 169 at r2](https://reviewable.io:443/reviews/se-edu/addressbook-level2/120#-K_9xrSEu87IuVn6d6yo:-K_9xrSEu87IuVn6d6yp:b-cguhre) ([raw file](https://github.com/se-edu/addressbook-level2/blob/dc2b2eeb0c79955f3a1e33ebb5db220809a94db5/src/seedu/addressbook/data/tag/UniqueTagList.java#L169)):*
> ```Java
>                         && this.internalList.equals(((UniqueTagList) other).internalList));
>     }
> ```

Why was this line deleted? Usually we leave a blank line (or two) between last method and the closing brace of the class.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/se-edu/addressbook-level2/120)*

**Comment by damithc** [comment]
@pyokagan any further inputs?

File: test/java/seedu/addressbook/commands/FindCommandTest.java [hunk 33]
```diff
@@ -0,0 +1,70 @@
...
+import seedu.addressbook.data.person.ReadOnlyPerson;
+import seedu.addressbook.util.TestUtil;
+import seedu.addressbook.util.TypicalPersons;
+
+public class FindCommandTest {
+
+    private AddressBook addressBook;
+    private TypicalPersons td;
+    
+    private final List<ReadOnlyPerson> EMPTY_LIST = Collections.emptyList();
```
Comment (pyokagan) [inline test/java/seedu/addressbook/commands/FindCommandTest.java]: Naming doesn't add a lot of value. I think you can just inline `Collections.emptyList()` directly.

File: test/java/seedu/addressbook/commands/FindCommandTest.java [hunk 34]
```diff
@@ -0,0 +1,70 @@
...
+    @Before 
+    public void setUp() {
+        this.addressBook = new TypicalPersons().getTypicalAddressBook();
+        this.td = new TypicalPersons();
+    }
+
+    @Test
+    public void execute() throws IllegalValueException {
+        //same word, same case: matched
+        assertFindCommandBehavior(new String[]{"Amy"}, TestUtil.createList(td.amy));
```
Comment (pyokagan) [inline test/java/seedu/addressbook/commands/FindCommandTest.java]: Use `Arrays.asList()` instead? I'm not sure why `TestUtil.createList()` exists, actually. @yamgent ?

File: test/java/seedu/addressbook/util/TypicalPersons.java [hunk 35]
```diff
@@ -0,0 +1,54 @@
...
+            bill =  new Person(new Name("Bill Clint"), new Phone("92229222", false), new Email("bc@gmail.com", false), 
+                    new Address("2 Clementi Road", false), new UniqueTagList());
+            candy =  new Person(new Name("Candy Destiny"), new Phone("93339333", false), new Email("cd@gmail.com", false), 
+                    new Address("3 Clementi Road", false), new UniqueTagList());
+        } catch (IllegalValueException e) {
+            e.printStackTrace();
+            assert false : "not possible";
+        }
+        
+
```
Comment (pyokagan) [inline test/java/seedu/addressbook/util/TypicalPersons.java]: Remove these empty lines.

File: test/java/seedu/addressbook/util/TypicalPersons.java [hunk 36]
```diff
@@ -0,0 +1,54 @@
...
+/**
+ * Class to generate typical test persons
+ */
+public class TypicalPersons {
+
+    public Person amy, bill, candy;
+    
+    public TypicalPersons() {
+        try {
+            amy =  new Person(new Name("Amy Buck"), new Phone("91119111", false), new Email("ab@gmail.com", false),
```
Comment (pyokagan) [inline test/java/seedu/addressbook/util/TypicalPersons.java]: One space between the `=` only.

File: test/java/seedu/addressbook/commands/FindCommandTest.java [hunk 37]
```diff
@@ -0,0 +1,70 @@
...
+import seedu.addressbook.data.AddressBook;
+import seedu.addressbook.data.exception.IllegalValueException;
+import seedu.addressbook.data.person.ReadOnlyPerson;
+import seedu.addressbook.util.TestUtil;
+import seedu.addressbook.util.TypicalPersons;
+
+public class FindCommandTest {
+
+    private AddressBook addressBook;
+    private TypicalPersons td;
```
Comment (pyokagan) [inline test/java/seedu/addressbook/commands/FindCommandTest.java]: Can make these final as well.

Actually, I think you can just assign the fields here and do without the `setUp()` method (JUnit constructs a new instance for every test).

**Comment by KnewYouWereTrouble** [comment]
Ready for review!

**Comment by pyokagan** [comment]
@KnewYouWereTrouble Just fix these problems below. Configure your IDE to automatically trim trailing whitespace so you won't have this problem any more. Everything else looks good.

```
[ant:checkstyle] [ERROR] /home/pyokagan/pyk/addressbook-level2/src/seedu/addressbook/data/tag/UniqueTagList.java:169: Line matches the illegal pattern 'Trailing whitespace'. [Regexp]
[ant:checkstyle] [ERROR] /home/pyokagan/pyk/addressbook-level2/test/java/seedu/addressbook/commands/FindCommandTest.java:35: Line matches the illegal pattern 'Trailing whitespace'. [Regexp]
[ant:checkstyle] [ERROR] /home/pyokagan/pyk/addressbook-level2/test/java/seedu/addressbook/commands/FindCommandTest.java:46: Line matches the illegal pattern 'Trailing whitespace'. [Regexp]
```

File: test/java/seedu/addressbook/commands/FindCommandTest.java (hunk 34 above)
Comment (yamgent) [inline test/java/seedu/addressbook/commands/FindCommandTest.java]: `createList()` is redundant, my bad. 

(I was under the impression that `Arrays.asList()` was returning `List<Person>` instead of `List<ReadOnlyPerson>`, because my parameters were all `Person`s. Looks like I was wrong.)

Comment (damithc) [inline test/java/seedu/addressbook/commands/FindCommandTest.java]: @yamgent create an issue to get rid of the unused method? We can leave it for new contributors.

Comment (yamgent) [inline test/java/seedu/addressbook/commands/FindCommandTest.java]: @damithc ok, created #134.

**Comment by KnewYouWereTrouble** [comment]
Ready for review!
